### PR TITLE
PML-266 Sparse tensor handling 

### PR DESCRIPTION
**Pull request by ben9871**
## Summary
Replaces whole-support materialization in the superposition path with chunked streaming accumulation to reduce peak memory usage.

## Related Issue
Related Jira: PML-266

## Type of change
- [ ] Bug fix
- [ ] New feature
- [ ] Documentation update
- [x] Refactor / Cleanup
- [x] Performance improvement
- [ ] CI / Build / Tooling
- [ ] Breaking change (requires migration notes)

## Proposed changes
- Introduces chunked superposition execution in `ComputationProcess`.
- Keeps `compute_ebs_simultaneously(...)` for batched simultaneous-process evaluation.
- Preserves sparse-input compatibility and numerical equivalence.
- Routes `QuantumLayer` through the new streamed path without changing the public API.

## How to test / How to run
1. Command lines

```bash
pytest tests/algorithms/test_amplitude_encoding.py tests/algorithms/test_compute_ebs_simultaneously.py tests/algorithms/test_superposition_state.py -q
ruff format --check merlin/algorithms/layer.py merlin/core/process.py tests/algorithms/test_amplitude_encoding.py tests/algorithms/test_compute_ebs_simultaneously.py tests/algorithms/test_superposition_state.py
ruff check merlin/algorithms/layer.py merlin/core/process.py tests/algorithms/test_amplitude_encoding.py tests/algorithms/test_compute_ebs_simultaneously.py tests/algorithms/test_superposition_state.py
mypy --ignore-missing-imports --no-strict-optional --allow-untyped-defs --disable-error-code=no-redef merlin/algorithms/layer.py merlin/core/process.py
```

## Screenshots / Logs (optional)
Benchmark evidence was captured locally during development, but is intentionally not committed.

Representative comparison on a 3-photon workload where both versions complete:

- workload: `m=16`, `n=3`, `ComputationSpace.FOCK`, `simultaneous_processes=16`
- compact support size: `C(18, 3) = 816`
- current chunked path: `0.760 s`, `739,772 KB` RSS
- old pre-chunked path: `14.209 s`, `1,447,772 KB` RSS

## Performance considerations (optional)
This branch is specifically about peak-memory reduction. The streaming path trades a small amount of orchestration overhead for bounded working-set size.

## Benchmark / test methodology
The benchmark comparison was run against the pre-chunked parent commit and the current chunked branch on the same workload and environment.

- Both versions were exercised on the same 3-photon case: `m=16`, `n=3`, `ComputationSpace.FOCK`, `simultaneous_processes=16`.
- The result measured end-to-end script runtime and process RSS for each version.

The compared workload was effectively:

```python
import os
import time
import resource

import torch
import perceval as pcvl

from merlin import ComputationSpace, MeasurementStrategy, QuantumLayer

m, n = 16, 3
layer = QuantumLayer(
    circuit=pcvl.Circuit(m),
    n_photons=n,
    measurement_strategy=MeasurementStrategy.amplitudes(
        computation_space=ComputationSpace.FOCK
    ),
)

amplitude_input = torch.zeros(816, dtype=torch.complex64)
amplitude_input[0] = 1.0 + 0.0j

start = time.perf_counter()
out = layer(amplitude_input, simultaneous_processes=16)
elapsed = time.perf_counter() - start
rss_kb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
print(elapsed, rss_kb, out.shape)
```

The same harness was run against the pre-chunked parent commit and the current branch.

Functional correctness was validated separately with:

```bash
pytest tests/algorithms/test_amplitude_encoding.py tests/algorithms/test_compute_ebs_simultaneously.py tests/algorithms/test_superposition_state.py -q
ruff check merlin/
ruff format --check merlin/
mypy --ignore-missing-imports --no-strict-optional --allow-untyped-defs --disable-error-code=no-redef merlin/algorithms/layer.py merlin/core/process.py
```

## Documentation
- [ ] User docs updated (Sphinx)
- [ ] Examples / notebooks updated
- [x] Docstrings updated
- [ ] Updated the API

## Checklist
- [x] PR title includes Jira issue key (e.g., PML-126)
- [x] "Related Jira ticket" section includes the Jira issue key (no URL)
- [x] Code formatted (ruff format)
- [x] Lint passes (ruff)
- [x] Static typing passes (mypy) if applicable
- [x] Unit tests added/updated (pytest)
- [x] Tests pass locally (pytest)
- [ ] Tests pass on GPU (pytest)
- [x] Test coverage not decreased significantly
- [ ] Docs build locally if affected (sphinx)
- [ ] With this command:

        > SPHINXOPTS="-W --keep-going -n" make -C docs clean html

    the docs are built without any warning or errors.
- [ ] New public classes/methods/packages are added in the API following the methodology presented in other files.
- [ ] Dependencies updated (if needed) and pinned appropriately
- [x] PR description explains what changed and how to validate it

### CODE REVIEW

**Comment by LF-Vigneux** [comment]
This point of the acceptance criteria is not present: Docs explain the chunked execution tradeoff

**Review by LF-Vigneux — CHANGES_REQUESTED** [review]
Except my previous comment everything seems to be working. My analysis of the tests and them pasing makes be believe that the code works as well. I am not fully grasping every detail of the implementation but in general everything seems to be good.

File: docs/source/user_guide/angle_amplitude_encoding.rst [hunk 1]
```diff
@@ -225,6 +225,42 @@ from the **type** of the first argument to ``forward()``:
...
+and converts dense tensors, dense :class:`~merlin.core.state_vector.StateVector`
+objects, sparse tensors, and sparse ``StateVector`` objects into the same
+internal active-support representation:
+
+.. code-block:: text
+
+   active basis indices + compact coefficient tensor [batch_size, nnz]
+
+Only those active Fock components are propagated through the simulator. Merlin
+processes them in chunks, calls the batched simulation kernel for each chunk,
```
Comment (LF-Vigneux) [inline docs/source/user_guide/angle_amplitude_encoding.rst]: "calls the batched simulation kernel" Is it really a kernel. Could be misinterpreted by some users.

File: merlin/core/process.py [hunk 2]
```diff
@@ -561,11 +493,193 @@ def _prepare_superposition_tensor(self) -> torch.Tensor:
...
+                device=unitary.device,
+                dtype=unitary.dtype,
+            )
+            _, batch_amplitudes = self.simulation_graph.compute_batch(
+                unitary, batch_fock_states
+            )
+            batch_amplitudes = batch_amplitudes / batch_amplitudes.norm(
+                p=2, dim=1, keepdim=True
+            ).clamp_min(1e-12)
+            final_amplitudes += torch.einsum("se,boe->bso", coeffs, batch_amplitudes)
```
Comment (LF-Vigneux) [inline merlin/core/process.py:459]: Here I dont understand the shapes, why is amplitudes three dimensional? I can understand coeffs being (batch_size,state_size) but I dont see the amplitudes dimension being 3D.

**Review by LF-Vigneux — COMMENTED** [review]
The code seems very sound, I did a deep dive of every new functions and everything seems to work. I understand what is going on and it seems to work. I openned two small discussions. I cannot find anything else to comment. 

As mentionned in our meeting, the documenttaion of the benchmark would be in another test.

So, after you checked my comments, you could remove the benchmark files and I will merge

File: merlin/core/process.py (hunk 2 above)
Comment (ben9871) [reply merlin/core/process.py:459]: batch_amplitudes has one extra dimension because compute_batch evaluates multiple input Fock states in the current chunk. Its shape is [parameter_batch, output_basis, chunk_size]. coeffs has shape [state_batch, chunk_size], so the einsum weights each simulated output vector by the corresponding input coefficient and sums over the chunk dim producing [parameter_batch, state_batch, output_basis]. For the common non-batched-parameter case, parameter_batch is just 1 and is squeezed later.

for every parameter batch b
for every state batch s
for every output state o

File: docs/source/user_guide/angle_amplitude_encoding.rst (hunk 1 above)
Comment (ben9871) [reply docs/source/user_guide/angle_amplitude_encoding.rst]: changed to simulation routine and pushed

**Review by LF-Vigneux — APPROVED** [review]
The code is sound and the documentation is clearly understandable.
